tests/test_cmake_config.py

```diff
@@ -1,28 +1,71 @@
 from __future__ import annotations
 
+import importlib
 import os
 import shutil
+import sysconfig
 from pathlib import Path
 from typing import TYPE_CHECKING
 
 import pytest
 from packaging.specifiers import SpecifierSet
 from packaging.version import Version
 
+from scikit_build_core import cmake
 from scikit_build_core.cmake import CMake, CMaker
 from scikit_build_core.errors import CMakeNotFoundError
 
 if TYPE_CHECKING:
     from collections.abc import Generator
 
+    from _pytest.mark.structures import ParameterSet
+
 DIR = Path(__file__).parent.resolve()
 
 
-def configure_args(config: CMaker, *, init: bool = False) -> Generator[str, None, None]:
+def get_cmake_configure_test_parameters() -> (
+    tuple[tuple[str, str, str], list[ParameterSet]]
+):
+    return ("platform_system", "generator", "single_config"), [
+        pytest.param("win", None, False, id="only_win_round"),
+        pytest.param("win", "Ninja", True, id="win_ninja_round"),
+        pytest.param("win", "Makefiles", True, id="win_makefiles_round"),
+        pytest.param("win", "Others", False, id="win_others_round"),
+        pytest.param("linux", None, True, id="only_linux_round"),
+        pytest.param("linux", "Ninja", True, id="linux_ninja_round"),
+    ]
+
+
+def configure_cmake_configure_test(
+    platform_system: str,
+    generator: str | None,
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    # CMaker::single_config is initialized with
+    # not sysconfig.get_platform().startswith("win")
+    monkeypatch.setattr(
+        sysconfig,
+        "get_platform",
+        lambda: platform_system,
+    )
```

Comment:
Instead of monkeypatching `get_platform`, could you enable tests depending on the platform that the test is run on? Probably make `get_cmake_configure_test_parameters` generate appropriate lists, though I think it would be better if the tests are still added, but as skipped, in order to have a record in the pytest

Reply:
thank you for the review,
let me know if I correctly captured in https://github.com/scikit-build/scikit-build-core/pull/847/commits/01072eef67405f087f5844c667586781ff4ee46c what you mean.

conda on linux/darwin:
```
...
tests/test_cmake_config.py::test_init_cache[only_win_round] SKIPPED (run
only on Windows)                                                         [ 10%]
tests/test_cmake_config.py::test_init_cache[win_ninja_round] SKIPPED     [ 10%]
tests/test_cmake_config.py::test_init_cache[win_makefiles_round] SKIPPED [ 11%]
tests/test_cmake_config.py::test_init_cache[win_others_round] SKIPPED    [ 11%]
tests/test_cmake_config.py::test_init_cache[only_linux_round] PASSED     [ 11%]
tests/test_cmake_config.py::test_init_cache[linux_ninja_round] PASSED    [ 12%]
tests/test_cmake_config.py::test_too_old PASSED                          [ 12%]
tests/test_cmake_config.py::test_cmake_args[only_win_round] SKIPPED (run
only on Windows)                                                         [ 12%]
tests/test_cmake_config.py::test_cmake_args[win_ninja_round] SKIPPED     [ 13%]
tests/test_cmake_config.py::test_cmake_args[win_makefiles_round] SKIPPED [ 13%]
tests/test_cmake_config.py::test_cmake_args[win_others_round] SKIPPED    [ 14%]
tests/test_cmake_config.py::test_cmake_args[only_linux_round] PASSED     [ 14%]
tests/test_cmake_config.py::test_cmake_args[linux_ninja_round] PASSED    [ 14%]
tests/test_cmake_config.py::test_cmake_paths[only_win_round] SKIPPED     [ 15%]
tests/test_cmake_config.py::test_cmake_paths[win_ninja_round] SKIPPED    [ 15%]
tests/test_cmake_config.py::test_cmake_paths[win_makefiles_round] SKIPPED [ 15%]
tests/test_cmake_config.py::test_cmake_paths[win_others_round] SKIPPED   [ 16%]
tests/test_cmake_config.py::test_cmake_paths[only_linux_round] PASSED    [ 16%]
tests/test_cmake_config.py::test_cmake_paths[linux_ninja_round] PASSED   [ 17%]
...
```

conda on windows
```
...
tests/test_cmake_config.py::test_init_cache[only_win_round] PASSED       [ 10%]
tests/test_cmake_config.py::test_init_cache[win_ninja_round] PASSED      [ 10%]
tests/test_cmake_config.py::test_init_cache[win_makefiles_round] PASSED  [ 11%]
tests/test_cmake_config.py::test_init_cache[win_others_round] PASSED     [ 11%]
tests/test_cmake_config.py::test_init_cache[only_linux_round] SKIPPED    [ 11%]
tests/test_cmake_config.py::test_init_cache[linux_ninja_round] SKIPPED   [ 12%]
tests/test_cmake_config.py::test_too_old PASSED                          [ 12%]
tests/test_cmake_config.py::test_cmake_args[only_win_round] PASSED       [ 13%]
tests/test_cmake_config.py::test_cmake_args[win_ninja_round] PASSED      [ 13%]
tests/test_cmake_config.py::test_cmake_args[win_makefiles_round] PASSED  [ 13%]
tests/test_cmake_config.py::test_cmake_args[win_others_round] PASSED     [ 14%]
tests/test_cmake_config.py::test_cmake_args[only_linux_round] SKIPPED    [ 14%]
tests/test_cmake_config.py::test_cmake_args[linux_ninja_round] SKIPPED   [ 15%]
tests/test_cmake_config.py::test_cmake_paths[only_win_round] PASSED      [ 15%]
tests/test_cmake_config.py::test_cmake_paths[win_ninja_round] PASSED     [ 15%]
tests/test_cmake_config.py::test_cmake_paths[win_makefiles_round] PASSED [ 16%]
tests/test_cmake_config.py::test_cmake_paths[win_others_round] PASSED    [ 16%]
tests/test_cmake_config.py::test_cmake_paths[only_linux_round] SKIPPED   [ 16%]
tests/test_cmake_config.py::test_cmake_paths[linux_ninja_round] SKIPPED  [ 17%]
...
```

tests run normally would also show the reason:
```
❯ pytest -vv -ra --showlocals tests/test_cmake_config.py
======================================================= test session starts =======================================================
platform darwin -- Python 3.10.14, pytest-7.4.4, pluggy-1.0.0 -- /Users/lorepirri/miniconda3/envs/test/bin/python
cachedir: .pytest_cache
installed packages of interest: exceptiongroup==1.2.0 packaging==24.1 pathspec==0.12.1 pip==24.0 pytest==7.4.4 pytest-subprocess==1.5.0 scikit_build_core==0.9.8 setuptools==69.5.1 tomli==2.0.1 virtualenv==20.26.1 wheel==0.43.0
sysconfig platform: macosx-11.1-arm64
rootdir: /Users/lorepirri/workspace/tmp-repos/scikit_build_core-0.9.8
configfile: pyproject.toml
plugins: subprocess-1.5.0
collected 20 items                                                                                                                

tests/test_cmake_config.py::test_init_cache[only_win_round] SKIPPED (run only on Windows)                                   [  5%]
tests/test_cmake_config.py::test_init_cache[win_ninja_round] SKIPPED (run only on Windows)                                  [ 10%]
tests/test_cmake_config.py::test_init_cache[win_makefiles_round] SKIPPED (run only on Windows)                              [ 15%]
tests/test_cmake_config.py::test_init_cache[win_others_round] SKIPPED (run only on Windows)                                 [ 20%]
tests/test_cmake_config.py::test_init_cache[only_linux_round] PASSED                                                        [ 25%]
tests/test_cmake_config.py::test_init_cache[linux_ninja_round] PASSED                                                       [ 30%]
tests/test_cmake_config.py::test_too_old PASSED                                                                             [ 35%]
tests/test_cmake_config.py::test_cmake_args[only_win_round] SKIPPED (run only on Windows)                                   [ 40%]
tests/test_cmake_config.py::test_cmake_args[win_ninja_round] SKIPPED (run only on Windows)                                  [ 45%]
tests/test_cmake_config.py::test_cmake_args[win_makefiles_round] SKIPPED (run only on Windows)                              [ 50%]
tests/test_cmake_config.py::test_cmake_args[win_others_round] SKIPPED (run only on Windows)                                 [ 55%]
tests/test_cmake_config.py::test_cmake_args[only_linux_round] PASSED                                                        [ 60%]
tests/test_cmake_config.py::test_cmake_args[linux_ninja_round] PASSED                                                       [ 65%]
tests/test_cmake_config.py::test_cmake_paths[only_win_round] SKIPPED (run only on Windows)                                  [ 70%]
tests/test_cmake_config.py::test_cmake_paths[win_ninja_round] SKIPPED (run only on Windows)                                 [ 75%]
tests/test_cmake_config.py::test_cmake_paths[win_makefiles_round] SKIPPED (run only on Windows)                             [ 80%]
tests/test_cmake_config.py::test_cmake_paths[win_others_round] SKIPPED (run only on Windows)                                [ 85%]
tests/test_cmake_config.py::test_cmake_paths[only_linux_round] PASSED                                                       [ 90%]
tests/test_cmake_config.py::test_cmake_paths[linux_ninja_round] PASSED                                                      [ 95%]
tests/test_cmake_config.py::test_get_cmake_via_envvar PASSED                                                                [100%]

===================================================== short test summary info =====================================================
SKIPPED [4] tests/test_cmake_config.py:69: run only on Windows
SKIPPED [4] tests/test_cmake_config.py:134: run only on Windows
SKIPPED [4] tests/test_cmake_config.py:173: run only on Windows
================================================== 8 passed, 12 skipped in 0.08s ==================================================

```

Reply:
Thanks indeed that looks great. Probably the naming could be improved, but my naming sense sucks. Let's see what feedback Henry has.